gobblin-compaction/src/main/java/org/apache/gobblin/compaction/verify/InputRecordCountHelper.java

```diff
@@ -87,31 +92,92 @@ public long calculateRecordCount (Collection<Path> paths) throws IOException {
 
   /**
    * Read record count from a specific directory.
-   * File name is {@link InputRecordCountHelper#RECORD_COUNT_FILE}
+   * File name is {@link InputRecordCountHelper#STATE_FILE}
    * @param fs  file system in use
    * @param dir directory where a record file will be read
    * @return record count
    */
   public static long readRecordCount (FileSystem fs, Path dir) throws IOException {
-    if (!fs.exists(new Path(dir, RECORD_COUNT_FILE))) {
-      return 0;
+
+    State state = new State();
+    if (fs.exists(new Path(dir, STATE_FILE))) {
+      try (FSDataInputStream inputStream = fs.open(new Path(dir, STATE_FILE))) {
+        state.readFields(inputStream);
+      }
     }
 
-    try (BufferedReader br = new BufferedReader(new InputStreamReader(fs.open (new Path (dir, RECORD_COUNT_FILE)), Charsets.UTF_8))) {
-      long count = Long.parseLong(br.readLine());
-      return count;
+    if (!state.contains(CompactionSlaEventHelper.RECORD_COUNT_TOTAL)) {
+      if (fs.exists(new Path (dir, RECORD_COUNT_FILE))){
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(fs.open (new Path (dir, RECORD_COUNT_FILE)), Charsets.UTF_8))) {
+          long count = Long.parseLong(br.readLine());
+          return count;
+        }
+      } else {
+        return 0;
+      }
+    } else {
+      return Long.parseLong(state.getProp(CompactionSlaEventHelper.RECORD_COUNT_TOTAL));
     }
   }
 
   /**
    * Write record count to a specific directory.
-   * File name is {@link InputRecordCountHelper#RECORD_COUNT_FILE}
+   * File name is {@link InputRecordCountHelper#STATE_FILE}
    * @param fs file system in use
    * @param dir directory where a record file will be saved
    */
   public static void writeRecordCount (FileSystem fs, Path dir, long count) throws IOException {
-    try (FSDataOutputStream outputFileStream = fs.create(new Path(dir, RECORD_COUNT_FILE))) {
-      outputFileStream.writeBytes(Long.toString(count));
+
+    State state = new State();
+    if (fs.exists(new Path(dir, STATE_FILE))) {
+      try (FSDataInputStream inputStream = fs.open(new Path(dir, STATE_FILE))) {
+        state.readFields(inputStream);
+      }
+    }
+
+    state.setProp(CompactionSlaEventHelper.RECORD_COUNT_TOTAL, count);
+    try (DataOutputStream dataOutputStream = new DataOutputStream(fs.create(new Path(dir, STATE_FILE)))) {
```

Comment:
Should write to temp file and swap to avoid a corrupt file if the process crashes before the data is fully persisted. Also, pull the writing logic to a utility function since it is called in multiple places and will be more complex once the temp file writing is added.